056-kafka-agent-controller-configuration.md

```diff
@@ -0,0 +1,99 @@
+> For this proposal we will not be supporting the dynamic reconfiguration of pure `controller` nodes.
+
+The best solution would have above limitations being addressed in the Kafka upstream project.
+The [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum+and+add+Controller+Registration) aims to address the first one, enabling a Kafka Admin client to connect directly to a `controller` node.
+Once connected, getting the configuration from the `controller` node is not currently part of the KIP so something to discuss in the upstream community (maybe raising it in the KIP-919 or opening a new one).
+The problem related to the `controller` only configuration properties is not taken into account yet.
+
+While the KRaft mode is already defined as "production ready" and the migration process from ZooKeeper-mode is getting improvements and stability, the Strimzi project has to keep the pace in order to provide the best experience to the users.
+Having the best support and fix the issues in the Kafka upstream project could take long time, maybe not landing on time in the next Kafka 3.6.0 release.
+For this reason, a temporary solution, built into the Strimzi project, will unlock the next chunk of work for supporting an automatic ZooKeeper-mode to KRaft migration within the Strimzi Cluster Operator.
+
+## Proposal
+
+### Exposing KRaft `controller` configuration
+
+During the Kafka start up script, the node configuration is generated and saved on the disk in the `/tmp/strimzi.properties` file.
+As soon as the Kafka Agent starts, it can read this file and load its content in memory in a JSON format.
+The top level JSON value is an object with all string-typed values (i.e. even if they are numbers).
+Following, an example of the JSON object bringing the `controller` configuration.
+
+```json
+{
+    "broker.id": "0",
+    "node.id": "0",
+    "process.roles": "controller",
+    "controller.listener.name": "CONTROL_PLANE-9090",
+    ...
+    ...
+}
+```
+
+Such a configuration is returned through the `/v1/node-configuration` HTTP endpoint.
+Because this file never changing once the broker/controller process has started, even if the node has its configuration dynamically updated, there is no need for the Kafka Agent to watch the file.
+It is enough to load its content on the Kafka Agent startup and keep it as cached in memory.
+The same content will be just returned every time the Kafka Roller is asking for it.
+
+If the Kafka Roller is talking to a previous version of the Kafka Agent where the `/v1/node-configuration` HTTP endpoint doesn't exist, it will get a `404 (NOT FOUND)` error.
+In this case, it will log an error message and fail the reconciliation.
```

Comment:
How will this information be secured? It contains possibly sensitive information such as passwords etc.

Reply:
The Kafka Agent endpoint is exposed via HTTPS as it's already for the broker state endpoint.

Reply:
https://github.com/strimzi/strimzi-kafka-operator/pull/8892 adds 8080 port for the probes.

Reply:
but only for the probes. The broker state is still on HTTPS. And the config endpoint will be the same.

Reply:
Well even from the code you linked you see ...

```java
server.setConnectors(new Connector[] {httpsConn, httpConn});
```

so Jetty is able to handle both HTTP and HTTPS at the same time and having different endpoints on them.
So again broker state on HTTPS and probes on HTTP.